clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -156,60 +156,138 @@ print_entry(trace_entry_t entry)
-    //
-    std::vector<struct expected_output> entries = {
+    //    and encoding entries, without any stop timestamp.
+    // 2. filter data and instruction address stream using a cache, with a
+    //    stop timestamp.
+    std::vector<struct test_case> entries = {
         // Trace shard header.
-        { { TRACE_TYPE_HEADER, 0, { 0x1 } }, { true, true } },
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VERSION, { 0x2 } }, { true, true } },
-        // Set OFFLINE_FILE_TYPE_ENCODINGS in the file type.
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_FILETYPE, { 0x200 } }, { true, true } },
-        { { TRACE_TYPE_THREAD, 0, { 0x4 } }, { true, true } },
-        { { TRACE_TYPE_PID, 0, { 0x5 } }, { true, true } },
+        { { TRACE_TYPE_HEADER, 0, { 0x1 } }, true, { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VERSION, { 0x2 } },
+          true,
+          { true, true } },
+        // File type, also modified by the record_filter based on the filtering
+        // done by the filter functions.
+        { { TRACE_TYPE_MARKER,
+            TRACE_MARKER_TYPE_FILETYPE,
+            { OFFLINE_FILE_TYPE_NO_OPTIMIZATIONS | OFFLINE_FILE_TYPE_ENCODINGS } },
+          true,
+          { false, false } },
+        { { TRACE_TYPE_MARKER,
+            TRACE_MARKER_TYPE_FILETYPE,
+            { OFFLINE_FILE_TYPE_NO_OPTIMIZATIONS | OFFLINE_FILE_TYPE_DFILTERED } },
+          false,
+          { true, false } },
+        { { TRACE_TYPE_MARKER,
+            TRACE_MARKER_TYPE_FILETYPE,
+            { OFFLINE_FILE_TYPE_NO_OPTIMIZATIONS | OFFLINE_FILE_TYPE_ENCODINGS |
+              OFFLINE_FILE_TYPE_DFILTERED | OFFLINE_FILE_TYPE_IFILTERED } },
+          false,
+          { false, true } },
+        { { TRACE_TYPE_THREAD, 0, { 0x4 } }, true, { true, true } },
+        { { TRACE_TYPE_PID, 0, { 0x5 } }, true, { true, true } },
         { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, { 0x6 } },
+          true,
           { true, true } },
+
         // Unit header.
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0x7 } }, { true, true } },
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0x8 } }, { true, true } },
-        { { TRACE_TYPE_INSTR, 4, { 0xaa00 } }, { true, true } },
-        { { TRACE_TYPE_WRITE, 4, { 0xaa80 } }, { true, true } },
-        // Unit header.
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0x9 } }, { false, true } },
-        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0xa } }, { false, true } },
-        { { TRACE_TYPE_WRITE, 4, { 0xaa90 } }, { false, false } },
-        { { TRACE_TYPE_ENCODING, 0, { 0 } }, { false, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 0x7 } },
+          true,
+          { true, true } },
+        { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0x8 } },
+          true,
+          { true, true } },
+        { { TRACE_TYPE_INSTR, 4, { 0xaa00 } }, true, { true, true } },
```

Comment:
We don't really need this to be a perfectly valid trace. I may have omitted a few entries which would otherwise be present in a collected trace, for test brevity. It currently should be good to verify the behaviors we want to verify though.